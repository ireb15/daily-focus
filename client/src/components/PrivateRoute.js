import React, { useEffect, useState } from "react";
import { Redirect, Route } from "react-router-dom";
import createPersistedState from "use-persisted-state";
import axios from "axios";

/**
 * This component masks the specified route to users that are not authenticated through
 * the firebase application and redirects them to the login screen
 * The API call based authentication process is adapted from:
 * https://gist.github.com/EduVencovsky/f8f6c275f42f7352571c92a59309e31d
 * @param component The page view that the user is trying to access
 * @param rest The rest of the component props
 */
export default function PrivateRoute({ component: Component, ...rest }) {
    const [authenticated, setAuthenticated] = useState(undefined);

    const useTokenState = createPersistedState("token");
    const [token] = useTokenState();

    useEffect(() => {
        async function verifyToken() {
            const result = await axios({
                method: "post",
                url: "http://localhost:9000/verifyToken",
                headers: { authorization: token },
            }).then((response) => {
                console.log(response);
                if (response.status === 200) {
                    return true;
                }
                return false;
            });
            console.log("result: " + result);
            return result;
        }
        setAuthenticated(verifyToken());
    }, [token]);

    if (authenticated === undefined) {
        return <>authenticating...</>;
    }

    return (
        <Route
            {...rest}
            render={(routeProps) =>
                authenticated ? <Component {...routeProps} /> : <Redirect to="/" />
            }
        />
    );
}
