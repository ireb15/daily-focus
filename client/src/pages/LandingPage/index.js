import React, { useState, Fragment } from "react";
import { Button, makeStyles, Snackbar, TextField } from "@material-ui/core";
import { useHistory, Link } from "react-router-dom";

import FocusLogo from "../../images/focus-logo.svg";
import styles from "./style.module.scss";

const useStyles = makeStyles({
    root: {
        "& .MuiOutlinedInput-root": {
            borderRadius: 15,
        },
        "& .MuiInputLabel-root": {
            fontSize: "1.25rem",
            color: "#7D7D7D",
            fontFamily: "'Ropa Sans', sans-serif",
        },
    },
    buttonRoot: {
        background: "#30A0F5",
        borderRadius: 50,
        padding: "4px 25px",
        width: "100%",
    },
    label: {
        textTransform: "capitalize",
        fontFamily: "'Ropa Sans', sans-serif",
        fontSize: "1.75rem",
    },
});

export default function LandingPage() {
    const classes = useStyles();
    const history = useHistory();
    const [signupMsgboxOpen, setSignupMsgboxOpen] = useState(false);
    const [signupMsg, setSignupMsg] = useState("Sign up failed, please try again later");
    const [email, setEmail] = useState("");
    const [password, setPassword] = useState("");

    const fetchSignup = (username, password) => {
        console.log(username);
        console.log(password);
        let signupSuccess = true;
        fetch("http://localhost:9000/signup", {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
            },
            body: JSON.stringify({
                email: username,
                password: password,
            }),
            referrerPolicy: "no-referrer",
        })
            .then((response) => {
                if (!response.ok) {
                    signupSuccess = false;
                }
                return response.json();
            })
            .then((data) => {
                // Signup unsuccessful
                if (!signupSuccess && data.message) {
                    setSignupMsg(data.message);
                    setSignupMsgboxOpen(true);
                    return Promise.reject("Server returned non-200 response");
                }

                // Signup successful
                console.log(data);
                setSignupMsg("Sign up succesful");
                setSignupMsgboxOpen(true);
                history.push("/login");
            })
            .catch((e) => {
                console.log(e);
            });
    };

    const submitSignup = (e) => {
        e.preventDefault();
        fetchSignup(email, password);
    };

    const handleSignupMsgboxClose = (event, reason) => {
        if (reason === "clickaway") {
            return;
        }
        setSignupMsgboxOpen(false);
    };

    return (
        <div className={styles.page}>
            <div className={styles.container}>
                <div className={styles.focusIntro}>
                    <img className={styles.focusLogo} src={FocusLogo} alt="FOCUS" />
                    <p>
                        Daily Focus is a dashboard web app that surfaces vital at-a-glance
                        information to enable you to know what you need to know and get on with your
                        day.
                    </p>
                </div>
                <div className={styles.signUpContainer}>
                    <h1>Create new account!</h1>
                    <form onSubmit={submitSignup}>
                        <TextField
                            style={{ paddingBottom: 20 }}
                            classes={{ root: classes.root }}
                            fullWidth
                            id="outlined-basic"
                            variant="outlined"
                            label="Email address"
                            value={email}
                            onInput={(e) => setEmail(e.target.value)}
                        />

                        <TextField
                            style={{ paddingBottom: 20 }}
                            classes={{ root: classes.root }}
                            fullWidth
                            id="outlined-basic"
                            variant="outlined"
                            label="Password"
                            value={password}
                            onInput={(e) => setPassword(e.target.value)}
                        />

                        <Button
                            variant="contained"
                            color="primary"
                            type="submit"
                            classes={{ root: classes.buttonRoot, label: classes.label }}
                        >
                            Sign Up
                        </Button>
                    </form>
                    <Link to="/login">Already have an account? Sign in</Link>
                </div>
            </div>
            <Snackbar
                anchorOrigin={{ vertical: "bottom", horizontal: "center" }}
                open={signupMsgboxOpen}
                autoHideDuration={5000}
                message={signupMsg}
                onClose={handleSignupMsgboxClose}
                action={
                    <Fragment>
                        <Button color="inherit" size="small" onClick={handleSignupMsgboxClose}>
                            CLOSE
                        </Button>
                    </Fragment>
                }
            />
        </div>
    );
}
