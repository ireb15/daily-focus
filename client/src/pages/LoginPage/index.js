import { Button, makeStyles, Paper, Snackbar, TextField } from "@material-ui/core";
import { useState, Fragment } from "react";
import createPersistedState from "use-persisted-state";
import { useHistory, Link } from "react-router-dom";

import { ReactComponent as FocusLogo } from "../../images/focus-logo-w-accent.svg";
import styles from "./style.module.scss";

const useTokenState = createPersistedState("token");

const useStyles = makeStyles({
    root: {
        "& .MuiOutlinedInput-root": {
            borderRadius: 15,
        },
        "& .MuiInputLabel-root": {
            fontSize: "1.25rem",
            color: "#7d7d7d",
            fontFamily: "'Ropa Sans', sans-serif",
        },
    },

    paper: {
        borderRadius: "20px",
        boxShadow: "0px 0px 31px 13px rgba(0, 0, 0, 0.1)",
        margin: "auto",
        padding: "32px 32px 0 32px",
        width: "480px",
    },

    button: {
        background: "#30a0f5",
        borderRadius: 50,
        color: "#ffffff",
        textTransform: "none",
        fontFamily: "'Ropa Sans', sans-serif",
        fontSize: "1.75rem",
        marginTop: "10px",
        width: "100%",
    },
});

const LoginPage = () => {
    const [email, setEmail] = useState("");
    const [password, setPassword] = useState("");
    const [loginErrorOpen, setloginErrorOpen] = useState(false);
    const [, setToken] = useTokenState();

    const classes = useStyles();
    const history = useHistory();

    const fetchLogin = (username, password) => {
        fetch("http://localhost:9000/login", {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
            },
            body: JSON.stringify({
                email: username,
                password: password,
            }),
            referrerPolicy: "no-referrer",
        })
            .then((response) => {
                if (!response.ok) {
                    setloginErrorOpen(true);
                    return Promise.reject("Server returned non-200 response");
                } else {
                    return response.json();
                }
            })
            .then((data) => {
                setToken("Bearer " + data.token);
                history.push("/home");
            })
            .catch((e) => {
                console.log(e);
            });
    };

    const submitLogin = (e) => {
        e.preventDefault();
        fetchLogin(email, password);
    };

    const handleLoginErrorClose = (event, reason) => {
        if (reason === "clickaway") {
            return;
        }
        setloginErrorOpen(false);
    };

    return (
        <div className={styles.container}>
            <Paper className={classes.paper}>
                <div className={styles.logInContainer}>
                    <div style={{ marginBottom: "24px" }}>
                        <FocusLogo width={380} fill="#454545" />
                    </div>
                    <form onSubmit={submitLogin}>
                        <TextField
                            style={{ margin: "10px 0px", width: "100%" }}
                            classes={{ root: classes.root }}
                            label="Email address"
                            variant="outlined"
                            value={email}
                            onInput={(e) => setEmail(e.target.value)}
                        />
                        <TextField
                            style={{ margin: "10px 0px", width: "100%" }}
                            classes={{ root: classes.root }}
                            label="Password"
                            type="password"
                            variant="outlined"
                            value={password}
                            onInput={(e) => setPassword(e.target.value)}
                        />
                        <Button
                            color="primary"
                            className={classes.button}
                            type="submit"
                            variant="contained"
                        >
                            Sign In
                        </Button>
                    </form>
                    <div style={{ padding: "32px" }}>
                        <Link to="/">Don't have an account? Sign up</Link>
                    </div>
                </div>
            </Paper>
            <Snackbar
                anchorOrigin={{ vertical: "bottom", horizontal: "center" }}
                open={loginErrorOpen}
                autoHideDuration={5000}
                message="Username and/or password is incorrect"
                onClose={handleLoginErrorClose}
                action={
                    <Fragment>
                        <Button color="inherit" size="small" onClick={handleLoginErrorClose}>
                            CLOSE
                        </Button>
                    </Fragment>
                }
            />
        </div>
    );
};

export default LoginPage;
