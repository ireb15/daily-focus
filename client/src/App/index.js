import React from "react";
import { BrowserRouter, Redirect, Route, Switch } from "react-router-dom";
import LandingPage from "../pages/LandingPage";
import LoginPage from "../pages/LoginPage";
import HomePage from "../pages/HomePage";
import PrivateRoute from "../components/PrivateRoute";
import "./style.css";

function App() {
    return (
        <BrowserRouter>
            <Switch>
                <Route exact path="/">
                    <Redirect to="/signup" />
                </Route>

                <Route exact path="/signup">
                    <LandingPage />
                </Route>

                <Route exact path="/login">
                    <LoginPage />
                </Route>

                <PrivateRoute component={HomePage} path="/home" exact />
            </Switch>
        </BrowserRouter>
    );
}

export default App;
