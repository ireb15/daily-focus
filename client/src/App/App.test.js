import React from "react";
import { shallow } from "enzyme";
import ShallowRenderer from "react-test-renderer/shallow";

import App from ".";
import LandingPage from "../pages/LandingPage";
import LoginPage from "../pages/LoginPage";
import HomePage from "../pages/HomePage";
import PrivateRoute from "../components/PrivateRoute";

let component;

beforeEach(() => {
    component = shallow(<App />);
});

test("it renders correctly", () => {
    const shallowRenderer = new ShallowRenderer();
    const snapshotComponent = shallowRenderer.render(<App />);
    expect(snapshotComponent).toMatchSnapshot();
});

test("it contains a landing page", () => {
    expect(component.find(LandingPage)).toHaveLength(1);
});

test("it contains a login page", () => {
    expect(component.find(LoginPage)).toHaveLength(1);
});

test("it contains a home page", () => {
    expect(component.find(PrivateRoute).prop("component")).toBe(HomePage);
});
