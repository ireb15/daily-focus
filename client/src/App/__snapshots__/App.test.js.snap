// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders correctly 1`] = `
<AuthProvider>
  <BrowserRouter>
    <Switch>
      <Route
        exact={true}
        path="/"
      >
        <Redirect
          to="/signup"
        />
      </Route>
      <Route
        exact={true}
        path="/signup"
      >
        <LandingPage />
      </Route>
      <Route
        exact={true}
        path="/login"
      >
        <LoginPage />
      </Route>
      <PrivateRoute
        component={[Function]}
        exact={true}
        path="/home"
      />
    </Switch>
  </BrowserRouter>
</AuthProvider>
`;
